### PR TITLE
Hash + Ladder Agent

### PR DESCRIPTION
**Pull request by jaseweston**

### CODE REVIEW

File: projects/params_vs_compute/hash_ladder/hash_ladder.py
Comment (spencerp) [inline projects/params_vs_compute/hash_ladder/hash_ladder.py:173]: nit; Uncomment or remove.

File: projects/params_vs_compute/hash_ladder/hash_ladder.py
Comment (spencerp) [inline projects/params_vs_compute/hash_ladder/hash_ladder.py:32]: We removed Apex, didn't we? If we need to keep this in this file for reproducibility that's fine, but I imagine it will also end up being removed eventually.

File: projects/params_vs_compute/hash_ladder/hash_ladder.py
Comment (spencerp) [inline projects/params_vs_compute/hash_ladder/hash_ladder.py:132]: Looks like many of these properties are the same as `TransformerDecoder` and you already call `super().init()`. Perhaps you only need the layer building code here.

I can add a `build_layers` function to `TransformerDecoder` in a subsequent PR for overriding purposes.

File: projects/params_vs_compute/hash_ladder/hash_ladder.py
Comment (spencerp) [inline projects/params_vs_compute/hash_ladder/hash_ladder.py:330]: Same here; it looks like we're only overloading `HashLayerFFN`? If so, you can just create your `TransformerDecoderLayer` like this without overriding:
```
TransformerDecoderLayer.with_components(feedforward=HashLayerFFN)
```

**Review by spencerp — APPROVED** [review]
There's some code that can be consolidated, but I didn't see any red flags. Accepting to unblock and I can do a cleanup/consolidation pass later.

File: projects/params_vs_compute/hash_ladder/hash_ladder.py
Comment (jaseweston) [reply projects/params_vs_compute/hash_ladder/hash_ladder.py]: done

File: projects/params_vs_compute/hash_ladder/hash_ladder.py
Comment (jaseweston) [reply projects/params_vs_compute/hash_ladder/hash_ladder.py]: removed

File: projects/params_vs_compute/hash_ladder/hash_ladder.py
Comment (jaseweston) [reply projects/params_vs_compute/hash_ladder/hash_ladder.py:118]: ok, i'll look at these later after this initial merge, thanks

File: projects/params_vs_compute/hash_ladder/hash_ladder.py
Comment (jaseweston) [reply projects/params_vs_compute/hash_ladder/hash_ladder.py:315]: ok, i'll look at these later after this initial merge, thanks

Comment (jaseweston) [reply projects/params_vs_compute/hash_ladder/hash_ladder.py:313]: (but i think i can't do that because i need to pass through the original inputs in the decoder layer)

**Pull request merged**
